pennylane/devices/qutrit_mixed/measure.py

```diff
@@ -0,0 +1,308 @@
+from pennylane.operation import Observable
+from pennylane.typing import TensorLike
+
+from .utils import (
+    get_einsum_mapping,
+    resquare_state,
+    get_num_wires,
+    get_new_state_einsum_indices,
+    QUDIT_DIM,
+)
+from .apply_operation import apply_operation
+
+alphabet_array = math.asarray(list(alphabet))
+
+
+def apply_observable_einsum(obs: Observable, state, is_state_batched: bool = False):
+    r"""Applies an observable to a density matrix rho, giving obs@state
+
+    Args:
+        obs (Operator): Operator to apply to the quantum state
+        state (array[complex]): Input quantum state
+        is_state_batched (bool): Boolean representing whether the state is batched or not
+
+    Returns:
+        TensorLike: the result of obs@state
+    """
+
+    def map_indices(
+        state_indices, row_indices, new_row_indices, **kwargs
+    ):  # pylint: disable=unused-argument
+        """map indices to wires
+        Args:
+            state_indices (str): Indices that are summed
+            row_indices (str): Indices that must be replaced with sums
+            new_row_indices (str): Tensor indices of the state
+            **kwargs (dict): Stores indices calculated in `get_einsum_mapping`
+
+        Returns:
+            String of einsum indices to complete einsum calculations
+        """
+        op_1_indices = f"{new_row_indices}{row_indices}"
+
+        new_state_indices = get_new_state_einsum_indices(
+            old_indices=row_indices,
+            new_indices=new_row_indices,
+            state_indices=state_indices,
+        )
+
+        return f"{op_1_indices},...{state_indices}->...{new_state_indices}"
+
+    num_ch_wires = len(obs.wires)
+    einsum_indices = get_einsum_mapping(obs, state, map_indices, is_state_batched)
+    obs_mat = obs.matrix()
+    obs_shape = [QUDIT_DIM] * num_ch_wires * 2
+    obs_mat = math.cast(math.reshape(obs_mat, obs_shape), complex)
+    return math.einsum(einsum_indices, obs_mat, state)
+
+
+def trace_method(
+    measurementprocess: ExpectationMP, state: TensorLike, is_state_batched: bool = False
+) -> TensorLike:
+    """Measure the expectation value of an observable by finding the trace of obs@rho.
+
+    Args:
+        measurementprocess (ExpectationMP): measurement process to apply to the state
+        state (TensorLike): the state to measure
+        is_state_batched (bool): whether the state is batched or not
+    Returns:
+        TensorLike: the result of the measurement
+    """
+    obs = measurementprocess.obs
+    rho_mult_obs = apply_observable_einsum(obs, state, is_state_batched)
+
+    # using einsum since trace function axis selection parameter names
+    # are not consistent across interfaces, they don't exist for torch
+
+    num_wires = get_num_wires(state, is_state_batched)
+    trace = math.einsum(f"...{alphabet[:num_wires]*2}", rho_mult_obs)
+    return math.real(trace)
+
+
+def reduce_density_matrix(  # TODO: ask if I should have state diagonalization gates?
```

Comment:
I'm unsure if I should have state diagonalization gates being applied here because it seems weird to have a measurement of the state with an observable

Reply:
Yeah it would be equivalent to getting the state represented in a different basis, I guess. I don't think the regular devices even have support for that.